src/ProxyClient/LiteSpeed.php

```diff
@@ -0,0 +1,124 @@
+<?php
+
+/*
+ * This file is part of the FOSHttpCache package.
+ *
+ * (c) FriendsOfSymfony <http://friendsofsymfony.github.com/>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace FOS\HttpCache\ProxyClient;
+
+use FOS\HttpCache\ProxyClient\Invalidation\ClearCapable;
+use FOS\HttpCache\ProxyClient\Invalidation\PurgeCapable;
+use FOS\HttpCache\ProxyClient\Invalidation\TagCapable;
+
+/**
+ * LiteSpeed Web Server (LSWS) invalidator.
+ *
+ * @author Yanick Witschi <yanick.witschi@terminal42.ch>
+ */
+class LiteSpeed extends HttpProxyClient implements PurgeCapable, TagCapable, ClearCapable
+{
+    private $headerLines = [];
+
+    /**
+     * {@inheritdoc}
+     */
+    public function clear()
+    {
+        $this->headerLines[] = 'X-LiteSpeed-Purge: *';
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function purge($url, array $headers = [])
+    {
+        $this->headerLines[] = 'X-LiteSpeed-Purge: '.$url;
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function configureOptions()
+    {
+        $resolver = parent::configureOptions();
+
+        $resolver->setRequired(['target_dir']);
+        $resolver->setAllowedTypes('target_dir', 'string');
+
+        return $resolver;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function invalidateTags(array $tags)
+    {
+        $this->headerLines[] = 'X-LiteSpeed-Purge: '.implode(', ', preg_filter('/^/', 'tag=', $tags));
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function flush()
+    {
+        $filename = $this->createFile();
+
+        $url = '/'.$filename;
```

Comment:
this is assuming we can put the files into the root directory. what if i need them in some sub directory?

lets rename target_dir to web_root to be clear what it is. right now, i'd expect /var/www/public/my_dir to work as target_dir, but it would end up with a 404

Reply:
Why? If you configure `https://example.com/my_dir` as base, this would work perfectly fine?

Reply:
ah, you are correct. can you do that in the example, specify the base url with a directory? and besides the example list the options and explicitly say how they interact? right now, the user has to be very familiar with FOSHttpCache or study the implementation to understand how this would work.

Reply:
Hmm, thinking about it, actually `$baseUri` never makes sense for LiteSpeed, am I right? We generate the file with the values we get. So if we get an `->purge(`/foobar`);` call, we generate `header('X-LiteSpeed-Purge: /foobar')` no matter the `$baseUri`. Wdyt?

Reply:
hm. we need to use the right base path in what we tell litespeed to purge. if the symfony app is installed at /foo and you purge /bar the web url is /foo/bar, right?

Reply:
I've checked with the LSWS support. So when you call e.g. `https://foobar.com/about-us` LS will of course cache this at `foobar.com/about-us`. If you want to purge this, you have to set the header `X-LiteSpeed-Purge: /about-us` on `foobar.com`. LSWS does not allow to purge cross-domain.
So `$baseUri` actually **must** be empty and `$servers` might be multiple ones if your site runs on multiple domains (which shouldn't be done anyway because of duplicate content).
We could ensure we only have the path in `public function purge($url...)`. So we might want to make absolute URLs relative ones automatically and document the part about having multiple servers and always leaving `$baseUri` empty. Do you understand what I mean? :)

Reply:
thats why i asked for $servers to be set to IPs to make the example clearer ;-)

so $baseUri is how your server is being talked to, including `foobar.com`. $servers is the list of IPs where the caching proxy runs. imho $baseUri must be the domain (plus path, if symfony does not run at `/`) and $server is the IP or DNS name of the machine running LSWS. 

multiple domain names for the same site are not supported with Varnish either. i don't think we want to support that, because it is extremly bad practice if you care about caching.

i think the target_dir, base_uri and server are all we need to specify things. we should explain how they work together with an example where symfony does not run at `/` of the domain, to avoid confusion there.

Reply:
> i think the target_dir, base_uri and server are all we need to specify things. we should explain how they work together with an example where symfony does not run at `/` of the domain, to avoid confusion there.

I'm trying but I don't get it myself. Imho `$baseUri` is prepended automatically to all relative `->purge($url)` calls. Is that not correct?

Reply:
it is prepended, yes. when the requests are actually sent, we use the $servers array to update the request: https://github.com/FriendsOfSymfony/FOSHttpCache/blob/a2fce472f38defcdb444d6c3c14257ff79c66392/src/ProxyClient/HttpDispatcher.php#L225-L233

the crucial line is https://github.com/FriendsOfSymfony/FOSHttpCache/blob/a2fce472f38defcdb444d6c3c14257ff79c66392/src/ProxyClient/HttpDispatcher.php#L231

in the end, HTTP uses a TCP connection to an IP and the payload contains a `Host: ` header. that host is what baseUri is about, and what the caching proxy looks at. the domain name in `$servers` is used to resolve the IP address (if its not directly an IP address already).

Reply:
Yeah, and we never prepend. Purging must be relative in our case (we can ensure this) and then servers must be the correct domains imho.

Reply:
imho we should follow the same logic as with the other caching proxies.

if symfony is not running at `/` we need to have the web path both in the path to the generated php file and prepend in the purge instructions inside that generated php file. i would use the base_uri for that path information. if we want to allow the generated php file to be created in a subpath, we would then need a separate parameter for that i think.

Reply:
ah sorry, how do you mean purging is relative? relative to the web path where symfony has its root?

Reply:
> I've checked with the LSWS support. So when you call e.g. https://foobar.com/about-us LS will of course cache this at foobar.com/about-us. If you want to purge this, you have to set the header X-LiteSpeed-Purge: /about-us on foobar.com. LSWS does not allow to purge cross-domain.

So `$purgeUri` is nonsense for LSWS. You cannot have `X-LiteSpeed-Purge: foobar.com/about-us`. You can only have it on the domain you request the generated purger file from.

Reply:
this is the same when purging from varnish. we create the request with the correct host name and the correct path: https://github.com/FriendsOfSymfony/FOSHttpCache/blob/b9d4bef48a678ee8b1b86391e9883c71e14d27b8/src/ProxyClient/HttpProxyClient.php#L93

i think to have the same feature scope, we should handle it when the $uri passed to `purge` contains the domain. so we would collect purge instructions per domain and on flush create those files and send the purge requests to $servers with the correct Host header for each domain.

Reply:
That doesn't work. I mean, for the urls, fine. But then to which file would you add the tags that need to be purged? All of them? Again, I don't think it works in a multi node setup so it never makes sense to provide multiple servers and so `$baseUri` is never needed.

Reply:
do you come to symfony con next week? i think it would be easier to discuss this face-to-face ;-)

Reply:
I'll be there by Wednesday at 3pm or so :)